pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -260,19 +258,25 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(env *model.Env
 			continue
 		}
 
-		for vsvcHost, gatewayHost := range matchingHosts {
-			if currentVhost, exists := vHostDedupMap[vsvcHost]; exists {
-				currentVhost.Routes = istio_route.CombineVHostRoutes(currentVhost.Routes, routes)
-			} else {
-				newVhost := &route.VirtualHost{
-					Name:    fmt.Sprintf("%s:%d", vsvcHost, port),
-					Domains: []string{vsvcHost, fmt.Sprintf("%s:%d", vsvcHost, port)},
-					Routes:  routes,
+		for vsvcHost, gatewayHosts := range matchingHosts {
+			for _, gatewayHost := range gatewayHosts {
+				host := vsvcHost
+				if gatewayHost.SubsetOf(model.Hostname(vsvcHost)) {
+					host = string(gatewayHost)
 				}
-				if tlsRedirect[gatewayHost] {
-					newVhost.RequireTls = route.VirtualHost_ALL
+				if currentVhost, exists := vHostDedupMap[host]; exists {
+					currentVhost.Routes = istio_route.CombineVHostRoutes(currentVhost.Routes, routes)
```

Comment:
So this is not correct.. imagine we have two gateway hosts (foo.com, bar.com) matching the same virtual service host (*.com).. We only have one set of http routes, but this loop runs twice for the same set of routes.. and it will end up appending the same set of routes twice to the existing vhost..

Reply:
No, it should create 2 different vhosts, one for foo.com and one for bar.com. Both will have the same set of routes, the ones that the VS defined for *.com.

Anyway, as we discussed offline, I agree that we should fix this code more generally, so I'll close this PR and work on another one that fixes the bug along with clean and proper logic for the whole thing.